### PR TITLE
Helper functions

### PR DESCRIPTION
**Pull request by faze-quadrata**
## Checklist
1. [ ] Unit Testing?
2. [ ] Backward Compatible?
3. [ ] Upgrade Safe (storage)?
4. [ ] Updated Confluence?
5. [ ] Updated Gitbook?

### CODE REVIEW

File: contracts/utility/QueryUtils.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >= 0.5.0;
+
+library QueryUtils {
+  // IsBusiness
+  function IsBusinessTrue(bytes32 _attrValue) public pure returns (bool) {
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:6]: docstring everywhere if possible

File: contracts/utility/QueryUtils.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  function AmlBetweenInclusive(bytes32 _attrValue, uint256 _lowerBound, uint256 _upperBound) public pure returns (bool) {
+    return(uint256(_attrValue) <= _upperBound && uint256(_attrValue) >= _lowerBound);
+  }
+
+  function AmlBetweenExclusive(bytes32 _attrValue, uint256 _lowerBound, uint256 _upperBound) public pure returns (bool) {
+    return(uint256(_attrValue) < _upperBound && uint256(_attrValue) > _lowerBound);
+  }
+
+  // Cred Protocol Score
+  function CredProtocolScoreIteratorLessThan(bytes32 _attrValue, bytes32 _startingHash, uint256 _iteratorThreshold) public pure returns (bool){
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:49]: You mean TU?

File: contracts/utility/QueryUtils.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    uint256 count = 0 ;
+    bytes32 hashIterator = _startingHash;
+
+    for(uint256 i = 0; i < 100; i++){
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:53]: 100 is a placeholder right? We wouldn't want to iterate 100 times to find out it's not enough?

File: contracts/utility/QueryUtils.sol (hunk 2 above)
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:49]: _startingHiddenScore

File: contracts/utility/QueryUtils.sol (hunk 1 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:7]: added

File: contracts/utility/QueryUtils.sol (hunk 2 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:79]: yes

Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:79]: Need to handle the case whre _attrValue == bytes32(0)

Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:79]: will update

File: contracts/utility/QueryUtils.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      if(hashIterator==_attrValue){
+        return(count <= _iteratorThreshold);
+      }
+      hashIterator = keccak256(abi.encodePacked(hashIterator));
+      count += 1;
+    }
+
+    return false;
+  }
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:92]: We're missing the CredProtocolScore helper

File: contracts/utility/QueryUtils.sol (hunk 3 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:83]: It depends on the number of tiers we'll have. Open to suggestions

Comment (fabrice-quadrata) [reply contracts/utility/QueryUtils.sol:83]: it should probably be total number of tiers +[2-5]? dunno might want to ping @travis-quadrata about his input

File: contracts/utility/QueryUtils.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >= 0.5.0;
+
+library QueryUtils {
+  // IsBusiness
+  function IsBusinessTrue(bytes32 _attrValue) public pure returns (bool) {
+    return(_attrValue == keccak256("TRUE"));
+  }
+
+  function IsBusinessFalse(bytes32 _attrValue) public pure returns (bool) {
```
[the remark this answers is not in the document]
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:13]: yes. fixed

File: contracts/utility/QueryUtils.sol (hunk 4 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:96]: added

File: contracts/tests/TestQueryUtils.sol [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    function IsBusinessTrue(bytes32 _attrValue) public pure returns(bool) {
+        return _attrValue.IsBusinessTrue();
+    }
+
+    function IsBusinessFalse(bytes32 _attrValue) public pure returns(bool) {
+        return _attrValue.IsBusinessFalse();
+    }
+
+    function CountryIsEqual(bytes32 _attrValue, string memory _expectedString) public pure returns(bool) {
```
Comment (teddy-quadrata) [inline contracts/tests/TestQueryUtils.sol:17]: Nit: Why are we starting function names with a capital letter? Seems to violate conventions no?

File: contracts/utility/QueryUtils.sol [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  /// @param _upperBound upper bound AML value as uint256
+  function AmlBetweenExclusive(bytes32 _attrValue, uint256 _lowerBound, uint256 _upperBound) public pure returns (bool) {
+    return(uint256(_attrValue) < _upperBound && uint256(_attrValue) > _lowerBound);
+  }
+
+  /// @dev Checks if supplied hash is within threshold of TU Credit Score value
+  /// @param _attrValue return value of query
+  /// @param _startingHiddenScore starting hidden score hash
+  /// @param _iteratorThreshold maximum number of hashing to meet criteria
+  function TuCreditScoreIteratorLessThan(bytes32 _attrValue, bytes32 _startingHiddenScore, uint256 _iteratorThreshold) public pure returns (bool){
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:79]: can we rename this VantageScore (to confirm with Springlabs) that this is the data being posted

File: contracts/utility/QueryUtils.sol [hunk 8]
```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >= 0.5.0;
+
+library QueryUtils {
```
Comment (fabrice-quadrata) [inline contracts/utility/QueryUtils.sol:4]: Is there a good way to handle upgrade for Libraries?

How are we gonna append new functions?

New deployments with versioning?

File: contracts/utility/QueryUtils.sol [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  /// @param _iteratorThreshold maximum number of hashing to meet criteria
+  function VantageScoreIteratorLessThan(bytes32 _attrValue, bytes32 _startingHiddenScore, uint256 _iteratorThreshold) public pure returns (bool){
+    if(_attrValue == bytes32(0)){
+      return false;
+    }
+
+    uint256 count = 0 ;
+    bytes32 hashIterator = _startingHiddenScore;
+
+    for(uint256 i = 0; i < 25; i++){
```
Comment (teddy-quadrata) [inline contracts/utility/QueryUtils.sol:87]: Hardcoding 25 won't work if our table is larger. We need to know the table length

File: contracts/utility/QueryUtils.sol [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+  /// @param _upperBound upper bound AML value as uint256
+  function AmlBetweenExclusive(bytes32 _attrValue, uint256 _lowerBound, uint256 _upperBound) public pure returns (bool) {
+    return(uint256(_attrValue) < _upperBound && uint256(_attrValue) > _lowerBound);
+  }
+
+  /// @dev Checks if supplied hash is within threshold of Vantage Score value
+  /// @param _attrValue return value of query
+  /// @param _startingHiddenScore starting hidden score hash
+  /// @param _iteratorThreshold maximum number of hashing to meet criteria
+  function VantageScoreIteratorLessThan(bytes32 _attrValue, bytes32 _startingHiddenScore, uint256 _iteratorThreshold) public pure returns (bool){
```
Comment (teddy-quadrata) [inline contracts/utility/QueryUtils.sol:79]: This looks to be doing a greater than check.

**Review by teddy-quadrata — CHANGES_REQUESTED** [review]

File: contracts/utility/QueryUtils.sol (hunk 10 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:79]: it's checking that the iterator counter is less than the bucket threshold count

Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:79]: the person calling this function isnt passing in the vantage score, they are passing in the number of max hashes

File: contracts/utility/QueryUtils.sol (hunk 9 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:87]: yes i've already pinged @travis-quadrata

File: contracts/tests/TestQueryUtils.sol (hunk 6 above)
Comment (faze-quadrata) [reply contracts/tests/TestQueryUtils.sol:17]: updated

File: contracts/utility/QueryUtils.sol (hunk 8 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol:4]: i'll look into this

File: contracts/utility/QueryUtils.sol (hunk 7 above)
Comment (faze-quadrata) [reply contracts/utility/QueryUtils.sol]: yes

**Review by teddy-quadrata — APPROVED** [review]

File: contracts/utility/QueryUtils.sol (hunk 8 above)
Comment (teddy-quadrata) [reply contracts/utility/QueryUtils.sol:4]: integrator logic contract will be linked with newest version of Quadrata/contracts npm package. no need for us to do anything special other than update npm package

File: contracts/utility/QueryUtils.sol (hunk 9 above)
Comment (travis-quadrata) [reply contracts/utility/QueryUtils.sol:87]: 25 should cover it. We are looking to keep bucket number <20. This gives us some head room, and we can also update if need be.

**Pull request merged**